### PR TITLE
Use constant-time comparisons

### PR DESCRIPTION
**Pull request by andyleiserson**
I suspect it would be quite difficult to construct a meaningful attack on one of these, but it still seems wise to use constant-time comparisons where best practice would dictate doing so.

### CODE REVIEW

File: ipa-core/src/protocol/basics/check_zero.rs [hunk 1]
```diff
@@ -61,7 +63,7 @@ where
             .expect("full reveal should always return a value"),
     );
 
-    Ok(rv == F::ZERO)
+    Ok(rv.ct_eq(&F::ZERO).into())
```
Comment (andyleiserson) [inline ipa-core/src/protocol/basics/check_zero.rs:66]: In `Fp32BitPrime` this isn't going to matter, but we're using MACs in `Fp25519` now.

File: ipa-core/src/ff/prime_field.rs [hunk 2]
```diff
@@ -241,6 +242,15 @@ macro_rules! field_impl {
             }
         }
 
+        // If the field value fits in a machine word, a naive comparison should be fine.
+        // But this impl is important for `[T]`, and useful to document where a
+        // constant-time compare is intended.
+        impl ConstantTimeEq for $field {
```
Comment (tyurek) [inline ipa-core/src/ff/prime_field.rs:248]: Do you think it makes sense to make the Eq implementation for the `field_impl` macro default constant time as well?

**Comment by akoshelev** [comment]
we should measure the performance impact of this change before merging. I agree constant time comparisons are important, but we need to understand whether this is going to make things significantly worse

**Comment by andyleiserson** [comment]
Draft run, 4M, before: https://draft-mpc.vercel.app/query/view/obese-oat2024-11-11T2252, 24:51
After: https://draft-mpc.vercel.app/query/view/gone-coke2024-11-11T2333, 24:48

**Review by akoshelev — APPROVED** [review]

**Comment by andyleiserson** [comment]
I am going to merge this. If we want to use the constant time algorithm for all equality checks, we can do that -- although there's nothing special about equality checks at that point -- it becomes a general exercise of scrubbing everything for constant-time execution. (Which it probably should be, but that is not a small undertaking.)

**Pull request merged**